Update x86 HWIntrinsic Tests

tannergooding opened the pull request:
I've updated the x86 hardware intrinsics to share some base code and to follow a template that ensures we get fairly consistent coverage.

Notable, The `TestTable` and `Program.cs` are shared.

All tests will be expected to have the following tests:
* Basic - Validates basic functionality works as intended
* Reflection - Validates invocation via reflection works as intended
* Unsupported - Validates PNSE is thrown when IsSupported == false
* Containment - Validates that functionality works as intended with various argument types that should be treated as contained
  * ClsVar - Static Variables
  * LclVar - Local Variables
  * LclFld - Field of a local variable
  * Fld - Field of the current instance
* Other - Validates other functionality that may be test specific (things like commutative or associativity tests can go here, if we need them)

At the very least, the new pattern is very "templatable" and is easy to copy and then do a search/replace on to get working for new tests

tannergooding commented:
FYI. @fiigii, @jkotas 

I'm not tied to this pattern (and it didn't take very long to update, as it was really just copy/paste and search/replace to get working, so if it is "throwaway" it's no big deal), but I think it might work well enough.

tannergooding commented:
Seems this found a bug under `_ro` as well.

For all the tests, the following pattern asserts:
```
var result = Sse.Add(
    Unsafe.Read<Vector128<float>>(_testTable.inArray1Ptr),
    Unsafe.Read<Vector128<float>>(_testTable.inArray2Ptr)
);
```

However, the following does not:
```
var left = Unsafe.Read<Vector128<float>>(_testTable.inArray1Ptr);
var right = Unsafe.Read<Vector128<float>>(_testTable.inArray2Ptr);

var result = Sse.Add(
    left,
    right
);
```

Assertion is:
```
Assert failure(PID 9644 [0x000025ac], Thread: 7680 [0x1e00]): Assertion failed 'argCnt == methInfo->args.totalILArgs()' in 'HardwareIntrinsicTest.Test:ValidateBasic():this' (IL size 67)
                                                                                                                            File: d:\repos\coreclr\src\jit\importer.cpp Line: 18195
```

tannergooding commented:
For reference: The above only happens for `COMPlus_EnableIncompleteISAClass=0` (so `Sse.IsSupported` returns false).

However, I have `Sse.IsSupported` indirected through `Test.IsSupported`, so the JIT doesn't constant fold the other calls away

tannergooding commented:
Looks like it might be a combination of inlining along with us returning a `gtNewMustThrowException` node for `IsSupported=false`: https://github.com/dotnet/coreclr/blob/master/src/jit/hwintrinsicxarch.cpp#L299.

4creators commented on tests/src/JIT/HardwareIntrinsics/X86/Shared/Program.cs:
```diff
@@ -0,0 +1,59 @@
+            {
+                var test = new Test();
+
+                if (test.IsSupported)
```
Split of tests into different calling patterns is very good, however, it could be more convenient to use Func<T> or similar to pass tests into main. This would allow for greater flexibility, since individual test would not have to implement all functions but only pass those which are relevant for them.

tannergooding replied:
The current functions represent the functionality that all tests should probably implement, in the most generic sense.

They ensure that the tests are functionally correct (for the various modes required) and that any new codegen works as intended.

There may be some tests which differ slightly, which is the purpose of the `ValidateOther`. For tests that differ more significantly, a more custom Main method is fine and would be expected.

4creators replied:
> The current functions represent the functionality that all tests should probably implement

Yes it is most probably correct but my intention was to suggest using more flexible pattern in discovering and running tests what would lift limitations on test implementation and decouple `Program` from `Test`. For instance every test method should implement `Func<bool>[] GetTests()` method which will return set of tests which will be run by `Main` function and some of the tests could be mandatory.

4creators replied:
And only `IsSupported` should be implemented directly by every test.

tannergooding replied:
> For instance every test method should implement Func<bool>[] GetTests() method

Fair enough. We should probably wait and see if moving to use a proper unit testing is viable, however (there may be some actual reason why not for CoreCLR)

4creators commented on tests/src/JIT/HardwareIntrinsics/X86/Shared/TestTable.cs:
```diff
@@ -0,0 +1,56 @@
+
+namespace HardwareIntrinsicTest
+{
+    public unsafe struct TestTable<T> : IDisposable where T : struct
```
TestTable in current form is very deficient for testing more complex operations. As a requirements I would propose:

1. Support all input data formats for all HW intrinsic operations - Vector128, Vector256, numerical types, an even strings for SSE42.
2. Store output of intrinsic under test as well as output of verification function alongside test data.
3. Allow for priniting to output of all collected test data for detailed verification.

Since we may need to have `TestTable` with 4 or even 5 data arrays it could be necessary to create set of `structs` or just switch to `class` and extend it via inheritance.

tannergooding replied:
I have some local work here that I haven't pushed yet.

This was basically renamed to `SimpleTestTable` and is meant for tests which take expect the same operation to be performed for each input. There are tests in other categories that would require their own new `TestTables`.

For example, some tests only operate on a single input, some operate on strings or scalar values, and some (like `Shuffle` or `Unpack`) require checking `inArray1`, `inArray2`, and `outArray` as a whole, rather than an element at time.

4creators replied:
I have some local work as well - I hope I can push it soon. It has all this difficult operations tested with several input - ouput combinations.

4creators commented on tests/src/JIT/HardwareIntrinsics/X86/Avx2/Add/UInt64.cs:
```diff
@@ -0,0 +1,149 @@
+        
+        private TestTable<ulong> _testTable;
+
+        public Test()
```
What I find very problematic here is the very limited test scope. Testing single operation on single Vector random elements just scratches the surface. In general Vector operations are fast and it is feasible to run 10s, 100s or even 1000s of them without any significant time penalty. 

Furthermore, we should test for all known problematic data points for given data format i.e. for floats we should have predefined problematic data comprising +0.0, -0.0, data close to minimum normal value, subnormal values, 1.0, -1.0 NaN, +infinity and -infinity, set of data around 0.0 and around 1.0 combined into operations with other large randomly generated data set.

Another requirement related to test function would be saving of all input data, output from function under test and output from verification function for later printout when test will fail to provide good diagnostic data.

All tests should run up to the end and only after that errors should be reported for all faulting operations.

tannergooding replied:
>  Testing single operation on single Vector random elements just scratches the surface.

These are meant to be unit tests, not integration or end-to-end tests. We should be writing other test types (integration, end-to-end, perf, etc) separately and only ensuring that the basic functionality is covered by these tests in particular.

> Furthermore, we should test for all known problematic data points for given data format i.e. for floats we should have predefined problematic data comprising 

I don't agree.

We are not coding complex algorithms, instead we are taking a method and (in the 95% of cases) generating a single hardware instruction. The vast majority of the complexity comes in the form of containment analysis and ensuring we do the right thing when hardware isn't supported or we are called indirectly.

Provided we are emitting the correct instruction and doing nothing else with the data (which is validated by the basic scenario) we can safely rely on the fact that Intel, AMD, and other hardware manufacturers have likely already done more extensive testing to ensure the underlying hardware instruction is functionally correct.

That beind said, there are few exceptions. These show up in the form of helper functions (Set, SetAll, etc) and for methods which require more extensive codegen (such as `CompareEqualUnorderedScalar`, which does a comparison, a couple of flag checks, and returns a bool).

> Another requirement related to test function would be saving of all input data, output from function under test and output from verification function for later printout when test will fail to provide good diagnostic data.

The way the current code is written, each test will print its failure information (which includes inputs and outputs) and all other tests will continue executing (short of an Exception being thrown or a JIT assert occurring).

4creators replied:
> we can safely rely on the fact that Intel, AMD, and other hardware manufacturers have likely already done more extensive testing to ensure the underlying hardware instruction is functionally correct

It is not my intention to test function correctness, however, trusting that HW manufacturers are doing their job well is OK but checking it is even better. My intention is to execute all corner cases which may or may not rise exceptions and test how we handle it. I am certain there could be some surprises ahead in there.

jkotas replied:
>  have likely already done more extensive testing 

This is not a safe assumption to make. You should assume that all testing that was done is using the tests committed in the repo.

tannergooding replied:
> You should assume that all testing that was done is using the tests committed in the repo.

My point was that we are providing a contract to the consumer of our API. That contract (for the majority of the hardware intrinsics) is the following:
* We will emit a particular instruction
* We will pass the inputs, as given
* We will return the output, as given

Our contract doesn't (and shouldn't) be giving any guarantees about the correctness of the underlying instruction.

We should be validating our contract (which is what the current tests are doing). Ensuring that special inputs for which the underlying instruction may behave differently is not something that our contract covers, as it is an implementation detail of the instruction we said we were going to emit.

4creators commented on tests/src/JIT/HardwareIntrinsics/X86/Avx2/Add/UInt64.cs:
```diff
@@ -0,0 +1,149 @@
+
+namespace HardwareIntrinsicTest
+{
+    public sealed unsafe class Test
```
I would prefer not to split operation tests into separate test for each operation - data type combination. This would really explode number of projects used to build and run the tests multiplying project files without significant gain in test content. From my checks building tests in coreclr consists from tiny actual code file build time and huge MSBuild overhead. It is even more pronounced during test run, where core logic test run is a minuscule portion of the whole test running time. 

Splitting larger tests into tiny ones will only exacerbate these problems.

tannergooding replied:
> I would prefer not to split operation tests into separate test for each operation - data type combination. This would really explode number of projects used to build and run the tests multiplying project files without significant gain in test content. From my checks building tests in coreclr consists from tiny actual code file build time and huge MSBuild overhead. It is even more pronounced during test run, where core logic test run is a minuscule portion of the whole test running time.

Ideally we would be using an actual unit test framework (such as xUnit or NUnit), there would be a project per ISA (rather than a project per method or per overload), a class per method or overload, and a Fact per scenario.

However, I've not seen many other tests in mscorlib (outside of perf tests) that support this.

It might be worth changing this however. @Jkotas, thoughts?

jkotas replied:
Unit test frameworks work great for testing framework APIs where it does not really matter how you call the API and it is fine to abstract it away using XUnit theories, etc.

They do not work well for testing codegen where you need the code calling the API to look in a particular way. (Also, the codegen/runtime bugs are pain to debug with unit testing framework loaded in the process since it introduces a lot of noise.)

I think you may want to look at autogenerating the test sources for testing the intrinsics.

tannergooding commented:
Cherry-picked the fix for the inliner assertion here: https://github.com/dotnet/coreclr/pull/15772

AndyAyersMS commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -292,12 +292,24 @@ GenTree* Compiler::impX86HWIntrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HA
+        // (gtIsRecursiveCall == false) as the gtNewMustThrowException node can mess with the
+        // inliner in unexpected ways.
+
+        if (gtIsRecursiveCall(method))
```
Getting back to my comment on your last PR, now that we have a use case -- I'd prefer it if this "recursive call ==> must expand" idiom was confined to the importer; you should pass `mustExpand` down in here to figure out whether bailing out is acceptable behavior.

tannergooding replied:
Will update.

tannergooding replied:
Fixed. Also see https://github.com/dotnet/coreclr/pull/15772, which has the changes standalone.

tannergooding commented:
Updated to use a T4 text template for generating the tests. Currently just a `SimpleBinaryOpTest` template which will cover a large number of the intrinsics.

We can add other templates based on test needs.

Working on a change to add JitLog checking to validate the calls are expanded as intrinsics where appropriate, but that can always go in another PR as well.

tannergooding commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

tannergooding commented:
The two job failures are unrelated to this change (which only touches HWIntrinsic test code). I've requeued the jobs.

Failure was: [baseservices_threading/_mutex_misc_waitone2_waitone2_](https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x64_checked_windows_nt_jitx86hwintrinsicnoavx_prtest/2/testReport/junit/baseservices_threading/_mutex_misc_waitone2_waitone2_/_mutex_misc_waitone2_waitone2_cmd/)

tannergooding commented:
@fiigii, @jkotas. Do these changes look good to you?

fiigii commented:
Question. Shall we extract `struct TestTable<T>` to be shared by all the tests?

tannergooding commented:
>  Shall we extract struct TestTable<T> to be shared by all the tests?

I think we can now that I have pulled the validation logic out of it (now it just represents three pinned arrays and should be reusable almost everywhere).

The place where it isn't still usable is for the scalar tests, which read/write a single value from memory.

tannergooding commented:
FYI. @CarolEidt as well.

4creators commented:
> I would prefer not to split operation tests into separate test for each operation - data type combination. This would really explode number of projects used to build and run the tests multiplying project files without significant gain in test content. From my checks building tests in coreclr consists from tiny actual code file build time and huge MSBuild overhead. It is even more pronounced during test run, where core logic test run is a minuscule portion of the whole test running time.

> Splitting larger tests into tiny ones will only exacerbate these problems.

Just digged out important issue which got folded by guthyb

4creators commented:
Another paleontological discovery hidden by github:

https://github.com/dotnet/coreclr/pull/15771#discussion_r160028787

tannergooding commented:
> #15771 (comment)

My response to this is still the same. The hardware intrinsic contract (for all non-helper intrinsics) is that we will:
* Throw a PNSE on unsupported hardware
* Otherwise, we will:
  * emit a particular instruction (using the VEX encoding, on hardware that supports it)
  * pass the input (or a reference to the input), as given by the user
  * return the output (or a reference to the output), as given back by the instruction

Any particulars about how the underlying hardware operates on any given data is an implementation detail of the hardware instruction and is not something we need to or should be testing.

Provided we are passing in the data unmodified, then it doesn't matter if it is QNaN, SNaN, 0, -0, +Inf, -Inf, denormal, rational, irrational, positive, negative, etc.

tannergooding commented:
It would be an entirely different story if we were providing a software fallback. But we are not (and even in the case of the compiler fallback, we are still meeting the original contract requirements stated).

fiigii commented:
This test pattern looks good to me. Is there anybody can approve and merge this PR?

4creators commented:
Do we care about test efficiency? If yes pls comment on my previous question.

fiigii commented:
> Do we care about test efficiency? 

I did not see any problem with the " test efficiency" yet.

tannergooding commented:
@fiigii, The "issue" @4creators is referring to is the large amount of time it takes to compile all the tests in the framework.

I currently have the tests split out into one project per method instance (so `Avx\Add\Single.cs`, `Avx\Add\Double.cs`, `Avx2\Add\Byte.cs`, etc). This means that there is some additional level of overhead to compile each project and/or to check they are up-to-date.

However, I don't know if this is a real concern. Once tests have been compiled, you can skip building tests again, and when modifying or adding new tests, it is very easy to build just the modified/new projects.

Additionally, I believe this split helps break the tests out into logical, readable, and debuggable components (and makes it much simpler to use with the T4 text templates).

tannergooding commented:
It might be possible to modify the template to be able to produce multiple outputs, in which case we would be back to a single project per method (i.e. just `Avx\Add_r.csproj`, `Avx\Add_ro.csproj` `Avx2\Add_r.csproj`, etc). We would still have a code file per `method instance`, but a general Main method which invoked them one at a time.

However, I'm not sure its worth the additional effort to do (and if it is, I think it can be on the backlog of "nice to have").

4creators commented:
> Once tests have been compiled, you can skip building tests again, and when modifying or adding new tests, it is very easy to build just the modified/new projects.

Yes it's true, but in general the efficiency I have measured is very poor, see for example:
https://github.com/Microsoft/msbuild/issues/2843
https://github.com/dotnet/project-system/issues/2789#issuecomment-355816589

However the most problematic is the overhead of running tests: on average we would increase number of independent test runs x 8 - 10. Currently my whole SSE2 test suite runs 1:20 s (after meltdwon and spectre patches), if we would have similar run times of other large ISAs it would x 5, and now if we multiply it roughly by x 8 - x 10 the running time would be equal to total priority 0 tests run time. 

> I think it can be on the backlog of "nice to have"

I agree with @tannergooding that we can leave on backlog now, but we should fix it not as nice to have but as soon as it will be reasonable or required.

4creators commented:
Ahh, forgot to write 1:20 s test run is for x86/x64 Debug build, 40 s x86/x64 Checked build

4creators commented:
> Any particulars about how the underlying hardware operates on any given data is an implementation detail of the hardware instruction and is not something we need to or should be testing.

> Provided we are passing in the data unmodified, then it doesn't matter if it is QNaN, SNaN, 0, -0, +Inf, -Inf, denormal, rational, irrational, positive, negative, etc.

Why IMO we should test it. There are several reasons:
1. large number of the above data should rise exceptions at the CPU level - we should know how they will translate to dotnet exceptions and control their type and information provided.
2. we may hit some Jit or VM corner cases for some of this values which are known to be problematic
3. Intel or Arm intrinsics despite their name are provided with dotnet, and due to that it is dotnet duty to check if they work correctly, it would be completely different story if we would allow pure assembly inside C#, than I would agree, that all we provide is pure instruction for which dotnet does not take responsibility for how it works once it was correctly emitted.
4. any problems with the hardware intrinsics will be first attributed to dotnet due to the fact that every user would expect that while implementing intrinsics we checked that they work correctly and if not we know what are the quirks (it's kind of new product, built on basis of the other product).
5. if this product is going to get Microsoft official support ... the above should be solved.

tannergooding commented:
> large number of the above data should rise exceptions at the CPU level - we should know how they will translate to dotnet exceptions and control their type and information provided.

The CLR explicitly disables these exceptions by default. Users can still enable them, but they need to do so through P/Invokes or other scenarios.

I don't believe doing so is a supported scenario and the behavior of a user doing such may be undefined.

> We may hit some Jit or VM corner cases for some of this values which are known to be problematic

Many of these data types and instructions are already being used today (For FEATURE_SIMD, or regular floating-point code). There should be no issue with reading/writing these data types (as far as something like `movaps` is concernced, its just 128-bits of raw data).

> Intel or Arm intrinsics despite their name are provided with dotnet, and due to that it is dotnet duty to check if they work correctly, it would be completely different story if we would allow pure assembly inside C#, than I would agree, that all we provide is pure instruction for which dotnet does not take responsibility for how it works once it was correctly emitted.

The feature is effectively typesafe inline assembly for a select number of hardware intrinsics that may help in performance driven scenarios. Explicit validation will be done for general-purpose APIs to ensure that inputs/outputs are behaving as expected.

For most APIs or operations, we are contracting that we will do something, but how it is done is implementation defined. However, for the hardware intrinsics, we are contracting that a particular instruction will be emitted, and are not contracting the functionality of said instruction.

> any problems with the hardware intrinsics will be first attributed to dotnet due to the fact that every user would expect that while implementing intrinsics we checked that they work correctly and if not we know what are the quirks (it's kind of new product, built on basis of the other product).

We are checking that they work correctly, by validating the basic functionality. However, any quirks of an instruction beyond that (outside of us not meeting our contract, such as by emitting bad instructions or encodings) are on the hardware and documentation of what they do should be deffered to the Intrinsics Guide for a given architecture (https://software.intel.com/sites/landingpage/IntrinsicsGuide/).

tannergooding commented:
To clarify, the Runtime disables the floating-point exceptions by default.

Something like `movaps` will raise an `AccessViolation` exception if you pass in an unaligned address. That is something we should validate in our tests.

jkotas commented:
> movaps will raise an AccessViolation

It would be better for this to be DataMisalignedException

tannergooding commented:
> It would be better for this to be DataMisalignedException

I agree. AccessViolation is just what the runtime throws today, with no special handling (this is likely just the default handling of `#GP(0)`)

4creators commented:
@tannergooding 

If we go further with your proposal regarding testing, than, we should at least in docs explicitly and very clear point to contract limitations, what is a contract which dotnet offers (emit instruction or helper function and handle correctly stack, register allocation and ... ??), settings of VM which impact instruction execution i.e. floating point exceptions,  memory alignment etc.

4creators commented:
> doing so is a supported scenario and the behavior of a user doing such may be undefined.

IEEE 754:2008 explicitly states that this will be supported

tannergooding commented:
The runtime spec (ECMA-335) is only IEC 60559:1989 compliant (which is identical, contentwise, to IEEE 754:1985).

The option of implementing exceptions is left up the the implementor in the original standard (IIRC).

Regardless, the ECMA spec explicitly does not do exception handling today. Supporting it is a different discussion and validating exceptions is something that would happen if such a change was agreed to be good (after going through the appropriate review process):

Current ECMA spec:
![image](https://user-images.githubusercontent.com/10487869/34900937-393fa48e-f7b9-11e7-901c-83168cce2feb.png)

4creators commented:
> The runtime spec (ECMA-335) is only IEC 60559:1989 compliant (which is identical, contentwise, to IEEE 754:1985).

Good point. It was good discussion to have @tannergooding

tannergooding commented:
@jkotas. Would you be able to review this or tag the appropriate people to do so?

jkotas commented:
These are JIT tests. The change should be reviewed by CodeGen team.

cc @RussKeldorph @BruceForstall @CarolEidt

tannergooding commented:
Moved a few of the new SSE tests to the T4 Text Template.

tannergooding commented:
And a few more of the SSE tests moved over to the templates.

The remaining ones I will probably do in a new PR, after this is merged.

tannergooding commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

tannergooding commented:
Test failures are being dealt with in https://github.com/dotnet/coreclr/pull/15901. They are fallout from a bad merge between two PRs that caused an assert to incorrectly fire on non-AVX machines.

tannergooding commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

fiigii commented:
test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

tannergooding commented:
Looks like the tests are failing on Ubuntu, possibly due to the different calling convention (as an initial best guess).

tannergooding commented:
Ubuntu failures are also fallout from fields being added to https://github.com/dotnet/coreclr/pull/15897
The new fields cause it to be detected as a `multireg` compatible struct.

@CarolEidt, do you know the general area where those checks are (I imagine we have handling to prevent `Vector<T>` from being treated as multireg, in the same way we had handling to prevent struct promotion).

CarolEidt commented:
I apologize for my delay in reviewing this.

First, re:
> @CarolEidt, do you know the general area where those checks are (I imagine we have handling to prevent Vector<T> from being treated as multireg, in the same way we had handling to prevent struct promotion).

We don't have explicit handling to prevent `Vector<T>` from being handled as multireg. Rather, the fact that it is variable-sized and has overlapping fields, make it ineligible. It's not clear to me why we would *not* want the other vector types to be recognized as multireg, though I can see that it might create problems. I'd like to understand those problems first before we dismiss it as a possibility.

More general review comments on these tests:
I like the overall approach with the coverage of the various operand types as well as testing the unsupported path - very nice!

That said, I strongly echo @4creators concerns about exploding the number of test files. While a developer doesn't need to build all the tests each time, the test build time is indeed a significant issue for the CI system and other full-build scenarios - @jashook can you confirm? Furthermore, I think that using a generic approach similar to that done for the JIT/SIMD tests would be an improvement, as it would avoid unnecessary duplication of test sources (there is a shared VectorUtil.cs file that has generic methods for getting random arrays and validating results). While these tests as proposed may be easy to generate, once they've been checked in the duplication becomes more of a maintenance burden.

Regarding the validation of the HW capabilities - I agree with @tannergooding that these are effectively "inline assembly" for dotnet, and therefore it is really not the burden of the HW intrinsics testing to ensure that the HW instructions operate correctly. This is in contrast to the `System.Numerics.Vectors` methods, which should be equivalent, operationally, to performing the operation on each element of the vector. (Note that the use of random values, rather than explicitly testing boundary conditions, led to missing the incorrect implementation of `Vector<float>.ConvertToUInt32()`, #15848.)

A few comments and questions:
- The methods called `Test` could perhaps be better named `TestSetup`, as they don't do any actual testing.
- The methods called `ValidateXXX` are not just validating, they are actually doing the testing as well. I might call them `TestXXX`.
- I am not a fan of the use of `var`. I find it easier to read code that is explicitly typed, but perhaps that's not a shared concern. @dotnet/jit-contrib - it might be nice if we had guidelines for things such as this.
- In the `ValidateResult` that takes arrays, why have you peeled off the first iteration?

tannergooding commented:
> It's not clear to me why we would not want the other vector types to be recognized as multireg

The System V ABI treats `__m128` (Vector128<T>) and `__m256` (Vector256<T>) specially. They are not aggregate values, but are automatically classified as `SSE` and `SSEUP`.

These types should always be passed in XMM/YMM registers, unless the available registers are full, in which case they should be passed on the stack.

> Furthermore, I think that using a generic approach similar to that done for the JIT/SIMD tests would be an improvement

I'll see if I can get a fix done for this. I think a good middle ground might be having a single project per ISA and having the T4 templates still generate a cs file per class (this should reduce build time, but keep the tests isolated from eachother, for easy debugging, management, etc).

> The methods called Test

Those are constructors, not methods. I would need to change the name of the class.

> The methods called ValidateXXX are not just validating, they are actually doing the testing as well. I might call them TestXXX.

Will update.

> I am not a fan of the use of var.

Happy to update to be explicit.

> In the ValidateResult that takes arrays, why have you peeled off the first iteration?

So that the template can be shared between `Packed` and `Scalar` intrinsics. The packed intrinsics perform the same operation on each field of the result, while the scalar intrinsics perform one operation on the first result and perform a different check on the remaining three.

4creators commented:
> I think a good middle ground might be having a single project per ISA 

@tannergooding @CarolEidt Single project per ISA would translate into single large test comprising all ISA methods. Therefore to test single method in case of large classes (SSE, SSE2, SSE4.1, AVX, AVX2 and in future AVX512) we would have to run several hundred tests at once. 

If we would decide to implement one method tests for two projects (r, ro) we would have around 2 400 tests for accepted ISAs with granularity which would allow to test methods independently. Number of tests would be comparable to other larger tests groups. This would speed up significantly development workflow comprising, coding, testing, debugging cycle and should not impact heavily CI time as it would increase by around 2 min of wall clock time on 4 core machines which are used right know in comparison to roughly 40 min for CI priority 0 Windows wall clock times.

> Note that the use of random values, rather than explicitly testing boundary conditions, led to missing the incorrect implementation of Vector<float>.ConvertToUInt32(), #15848.

I am fully aware of this problem and this is why I have proposed to test explicitly for known "problematic" values

Actually, one the conclusions from the previous test plan discussion is that we should test "contract" which we provide to users. If we look carefully into scope of the abstraction which is provided by intrinsics it appears that we do something more than give a chance to write "inline assembly". We **explicitly relieve coders** from the following requirements of assembly programming:

1. Explicit register management and allocation.
2. Stack management.
3. Memory management.
4. Call management.

To some extent we share with coders responsibility for managing data transfer between `xmm`, `ymm` register and memory, depending on the method of data transfer chosen by coder.

IMO we should carefully consider how we can test for these functionality of Jit which is created right now and most probably cannot be tested with kind of simple tests we have written so far.

sdmaclea commented:
>I think a good middle ground might be having a single project per ISA

👍 

Since, we should be testing that the correct assembly instruction(s) is generated for each form, we should not be trying to extensively test the generated instruction.  A few random and/or fixed data sets should be sufficient.

As such, the tests are very brief and will be dominated by startup time / JIT time.  My ARM64 plan is one tests per ISA.  It should save substantial time.  Of course there is no reason this has to be a hard fixed rule.

sdmaclea commented:
> that we should test "contract" which we provide to users.

This is probably an orthogonal set of targeted tests.  ISA tests would provide an exhaustive set of intrinsic unit tests.  Contract tests would test other aspects.

tannergooding commented:
> Single project per ISA would translate into single large test comprising all ISA methods. Therefore to test single method in case of large classes (SSE, SSE2, SSE4.1, AVX, AVX2 and in future AVX512) we would have to run several hundred tests at once.

@4creators, we have existing test cases which do something similar, but they have logic in the `Main` method to support testing specific tests or specific subsets of tests. Ex: https://github.com/dotnet/coreclr/blob/master/tests/src/JIT/Performance/CodeQuality/Math/Functions/Functions.cs#L22

This removes the need to test everything when you are only modifying a single test.

4creators commented:
> but they have logic in the Main method to support testing specific tests or specific subsets of tests

@tannergooding  Agree, this is best of all worlds 👍

CarolEidt commented:
>> Note that the use of random values, rather than explicitly testing boundary conditions, led to missing the incorrect implementation of Vector.ConvertToUInt32(), #15848.

> I am fully aware of this problem and this is why I have proposed to test explicitly for known "problematic" values

I think you missed my point, which is that for `Vector<T>` which is an *abstraction*, it is incumbent on us to ensure its behavior matches scalar operations. However, the intrinsics are specifically *not* abstractions, so we shouldn't waste our efforts validating the functionality of the hardware, but rather focus on validating the interface - i.e. ensuring that operands and results are handled correctly.

tannergooding commented:
Ubuntu failures are being dealt with here: https://github.com/dotnet/coreclr/pull/15942

tannergooding commented:
Should have the changes up sometime tonight or tomorrow.

The T4 templates were not cooperating for multi-file generation (you can make it work, but it may not be cross-platform and has other issues as well).

I've switched it to have a `GenerateTests.csx` script (C# script) and a `SimpleBinaryOpTest.template` instead and am just polishing up a few things before I push.

tannergooding commented:
Ok, new changes are up.

Overall the tests converted are the same, but now it uses the `Shared\GenerateTests.csx` script and they share a single project file (with a name matching the ISA) rather than one project per test.

tannergooding commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

fiigii commented:
> but now it uses the Shared\GenerateTests.csx

Thanks for the work. So, is this script executed by the coreclr test build to automatically generate test cases? Or we need to manually run it?

Can we push this PR to get merged as soon as possible? There are more and more intrinsic PRs coming, so we should have a criteriron for HW intrinsic tests to avoid refactoring tests again and again...

tannergooding commented:
> Or we need to manually run it?

At present, it needs to be manually run after you update it with any new input data.

fiigii commented:
> it needs to be manually run after you update it with any new input data.

Good to know. Current test-template looks pretty good, and I think we can merge this PR at first, and extending the template later to support more tests (e.g., JIT-log) is easy.

cc @CarolEidt

sdmaclea commented on tests/src/JIT/HardwareIntrinsics/X86/Shared/SimpleBinOpTest.template:
```diff
@@ -0,0 +1,209 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
```
The all generated tests (and therefore the template) should contain a comment noting that the test is generated and should not be manually edited.  It should also contain instructions on how to regenerate.

CarolEidt commented:
@jashook - could you have a look at this? I believe that the use of a single project file per ISA avoids the excessive test build time that I was concerned about, but I'd like your input on this.

This looks good to me overall, but I think there is a need to add some minimal documentation. This is all pretty self-explanatory, I think, but there needs to be a way to guide a JIT/coreclr developer to find the right files. Probably a brief document under https://github.com/dotnet/coreclr/tree/master/Documentation/project-docs?

tannergooding commented:
Updated the generated files to indicate that they are automatically generated from the script and updated the script with basic instructions for how to run or add new tests.

fiigii commented on tests/src/JIT/HardwareIntrinsics/X86/Shared/SimpleBinOpTest.template:
```diff
@@ -0,0 +1,216 @@
+            }}
+            else
+            {{
+                for (var i = 1; i < left.Length; i++)
```
I think we can remove this `for` loop to enable the template for more intrinsics (e.g., `HorizontalAdd`).

tannergooding replied:
You mean just check each element individually, rather than looping?

fiigii replied:
Yes, that would make the template more general.

tannergooding replied:
I'm going to log an issue to track this (as part of porting the remaining tests to use the template as well).

We probably want to refactor it such that we don't have to duplicate a bunch of table data for the regular intrinsics (where each check will be the same).

tannergooding commented:
Windows_NT x86 Release Innerloop Build and Test passed, but hasn't updated the icon.

CarolEidt commented:
ping @jashook - do you think you'll find some time to review this soon?

jashook commented:
@CarolEidt sorry missed this. Taking a look now.

CarolEidt commented:
> It is worth noting the new tests will have to be added to our lstFile to run them in automation. My guess is that file is quite out of date already.

So, is it the case that newly-added tests are never run in automation? Does that mean ci, or ...?

tannergooding commented:
>  Does that mean ci

I'm fairly certain they are being run in CI, it is certainly failing the builds in a number of cases 😄

jashook commented:
That is the case only for arm/arm64.

tannergooding commented:
Logged: https://github.com/dotnet/coreclr/issues/16113

CarolEidt commented:
Thanks @jashook !
